File: app/src/main/java/io_developers/sssemil/com/wand/DeviceStatusView.java
package io_developers.sssemil.com.wand;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import static io_developers.sssemil.com.wand.MainActivity.BT_DEVICE_NAME;

/**
 * Created by emil on 8/05/17.
 */
public class DeviceStatusView extends RelativeLayout {

    public static final byte ERROR = -1;
    public static final byte DISCONNECTED = 0;
    public static final byte CONNECTING = 1;
    public static final byte CONNECTED = 2;

    private int mConnectionStatus;
    private TextView mTextView;
    private ImageView mIconView;

    private final BroadcastReceiver mReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            String action = intent.getAction();
            BluetoothDevice device = intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE);

            if (BluetoothDevice.ACTION_ACL_CONNECTED.equals(action)) {
                //Device is now connected
                if (device != null && device.getName().equals(BT_DEVICE_NAME)) {
                    setStatus(CONNECTED);
                }
            } else if (BluetoothDevice.ACTION_ACL_DISCONNECTED.equals(action)) {
                //Device has disconnected
                if (device != null && device.getName().equals(BT_DEVICE_NAME)) {
                    setStatus(DISCONNECTED);
                }
            }
        }
    };
    private IntentFilter mFilter;
    private boolean mRegisteredReceiver = false;

    public DeviceStatusView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    public DeviceStatusView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(context, attrs);
    }

    public void setStatus(int status) {
        mConnectionStatus = status;
        switch (mConnectionStatus) {
            case DISCONNECTED:
                mTextView.setText(R.string.disconnected);
                mIconView.setBackgroundResource(R.drawable.ic_disconnected);
                break;
            case CONNECTING:
                mTextView.setText(R.string.connecting);
                mIconView.setBackgroundResource(R.drawable.ic_connecting);
                break;
            case CONNECTED:
                mTextView.setText(R.string.connected);
                mIconView.setBackgroundResource(R.drawable.ic_connected);
                break;
            case ERROR:
                mTextView.setText(R.string.error);
                mIconView.setBackgroundResource(R.drawable.ic_disconnected);
                break;

        }
        invalidate();
        requestLayout();
    }

    private void init(Context context, AttributeSet attrs) {
        mFilter = new IntentFilter();
        mFilter.addAction(BluetoothDevice.ACTION_ACL_CONNECTED);
        mFilter.addAction(BluetoothDevice.ACTION_ACL_DISCONNECT_REQUESTED);
        mFilter.addAction(BluetoothDevice.ACTION_ACL_DISCONNECTED);

        registerReceiver();

        TypedArray a = context.getTheme().obtainStyledAttributes(
                attrs,
                R.styleable.DeviceStatusView,
                0, 0);

        try {
            mConnectionStatus = a.getInteger(R.styleable.DeviceStatusView_status, DISCONNECTED);
        } finally {
            a.recycle();
        }

        inflate(getContext(), R.layout.device_status_layout, this);
        this.mTextView = (TextView) findViewById(R.id.text);
        this.mIconView = (ImageView) findViewById(R.id.icon);

        setStatus(mConnectionStatus);
    }

    void unregisterReceiver() {
        if (mRegisteredReceiver) {
            getContext().unregisterReceiver(mReceiver);
            mRegisteredReceiver = false;
        }
    }

    void registerReceiver() {
        if (!mRegisteredReceiver) {
            getContext().registerReceiver(mReceiver, mFilter);
            mRegisteredReceiver = true;
        }
    }
}

File: app/src/main/java/io_developers/sssemil/com/wand/DeviceStatusView.kt
package io_developers.sssemil.com.wand

import android.bluetooth.BluetoothDevice
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.util.AttributeSet
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

/**
 * Created by emil on 8/05/17.
 */
class DeviceStatusView : LinearLayout {

    private var connectionStatus: Int = 0
    private var textView: TextView? = null
    private var iconView: ImageView? = null

    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action
            val device = intent.getParcelableExtra<BluetoothDevice>(BluetoothDevice.EXTRA_DEVICE)

            if (BluetoothDevice.ACTION_ACL_CONNECTED == action) {
                //Device is now connected
                if (device != null && device.name == MainActivity.BT_DEVICE_NAME) {
                    setStatus(CONNECTED.toInt())
                }
            } else if (BluetoothDevice.ACTION_ACL_DISCONNECTED == action) {
                //Device has disconnected
                if (device != null && device.name == MainActivity.BT_DEVICE_NAME) {
                    setStatus(DISCONNECTED.toInt())
                }
            }
        }
    }
    private var mFilter: IntentFilter? = null
    private var mRegisteredReceiver = false

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init(context, attrs)
    }

    fun setStatus(status: Int) {
        connectionStatus = status
        when (connectionStatus) {
            DISCONNECTED -> {
                textView!!.setText(R.string.disconnected)
                iconView!!.setBackgroundResource(R.drawable.ic_disconnected)
            }
            CONNECTING -> {
                textView!!.setText(R.string.connecting)
                iconView!!.setBackgroundResource(R.drawable.ic_connecting)
            }
            CONNECTED -> {
                textView!!.setText(R.string.connected)
                iconView!!.setBackgroundResource(R.drawable.ic_connected)
            }
            ERROR -> {
                textView!!.setText(R.string.error)
                iconView!!.setBackgroundResource(R.drawable.ic_disconnected)
            }
        }
        invalidate()
        requestLayout()
    }

    private fun init(context: Context, attrs: AttributeSet) {
        mFilter = IntentFilter()
        mFilter!!.addAction(BluetoothDevice.ACTION_ACL_CONNECTED)
        mFilter!!.addAction(BluetoothDevice.ACTION_ACL_DISCONNECT_REQUESTED)
        mFilter!!.addAction(BluetoothDevice.ACTION_ACL_DISCONNECTED)

        registerReceiver()

        val a = context.theme.obtainStyledAttributes(
                attrs,
                R.styleable.DeviceStatusView,
                0, 0)

        try {
            connectionStatus = a.getInteger(R.styleable.DeviceStatusView_status, DISCONNECTED.toInt())
        } finally {
            a.recycle()
        }

        View.inflate(getContext(), R.layout.device_status_layout, this)

        this.iconView = findViewById(R.id.icon) as ImageView
        this.textView = findViewById(R.id.text) as TextView

        setStatus(connectionStatus)
    }

    internal fun unregisterReceiver() {
        if (mRegisteredReceiver) {
            context.unregisterReceiver(receiver)
            mRegisteredReceiver = false
        }
    }

    internal fun registerReceiver() {
        if (!mRegisteredReceiver) {
            context.registerReceiver(receiver, mFilter)
            mRegisteredReceiver = true
        }
    }

    companion object {

        val ERROR: Int = -1
        val DISCONNECTED: Int = 0
        val CONNECTING: Int = 1
        val CONNECTED: Int = 2
    }
}
